### PR TITLE
Eliminate double autoscaling by fixing issued job code (resolves #1086)

### LINKED ISSUE #1086: Auto-scaling creates twice as many nodes as needed
… because non-preemptable nodes are scaled up even though all jobs are preemptable.

### PR DESCRIPTION
**Pull request by hannes-ucsc**
Resolves #1086
Continues #1112

### CODE REVIEW

File: src/toil/leader.py [hunk 1]
```diff
@@ -190,6 +202,8 @@ def removeJobID(self, jobBatchSystemID):
         """
         assert jobBatchSystemID in self.jobBatchSystemIDToIssuedJob
         self.jobsIssued -= 1
+        if self.jobBatchSystemIDToIssuedJob[jobBatchSystemID].preemptable:
+            self._preemptableJobsIssued -= 1
```
Comment (fnothaft) [inline src/toil/leader.py]: May want to assert here that `self._preemptableJobsIssued` doesn't go negative.

File: src/toil/provisioners/clusterScaler.py [hunk 2]
```diff
@@ -28,6 +28,17 @@
...
+# A *deficit* exists when we have more jobs that can run on preemptable nodes than we have
+# preemptable nodes. In order to not block these jobs, we want to increase the number of non-
+# preemptable nodes that we have and need for just non-preemptable jobs. However, we may still
+# prefer waiting for preemptable instances to come available.
+#
+# To accomodate this, we set the delta to the difference between the number of provisioned
+# preemptable nodes and the number of nodes that were requested. when the non-preemptable thread
+# wants to provision nodes, it will multiply this delta times a preference for preemptable vs.
+# non-preemptable nodes.
+
```
Comment (fnothaft) [inline src/toil/provisioners/clusterScaler.py]: Nit: extra whitespace?

File: src/toil/provisioners/clusterScaler.py [hunk 3]
```diff
@@ -331,6 +356,19 @@ def tryRun(self):
...
                                                                       preemptable=self.preemptable)
+                    
+                    # If we were scaling up the number of preemptable nodes and failed to meet
+                    # our target, we need to update the slack so that non-preemptable nodes will
+                    # be allocated instead and we won't block. If we _did_ meet our target,
+                    # we need to reset the slack to 0.
+                    if self.preemptable:
+                        if totalNodes < estimatedNodes:
+                            deficit = estimatedNodes - totalNodes
+                            logger.debug('Preemptable scaler detected deficit of %d.', deficit)
```
Comment (fnothaft) [inline src/toil/provisioners/clusterScaler.py]: I would probably make this message more descriptive and promote it to `logger.info`. IMO, it is useful to the end user to know that their bid price was insufficient to get preemptable nodes (or, that there wasn't sufficient capacity). Ignore this if we're logging something similar elsewhere; we may be.

File: src/toil/provisioners/clusterScaler.py (hunk 2 above)
Comment (hannes-ucsc) [inline src/toil/provisioners/clusterScaler.py]: Actually, ignore last comment.

Can you clarify what you mean. Unsure if you mean blank lines or sth else.

Comment (hannes-ucsc) [inline src/toil/provisioners/clusterScaler.py]: There was a blank line missing after the initialization of `_preemptableNodeDeficit` and I fixed that.

**Comment by fnothaft** [comment]
@hannes-ucsc just realized, you have commits that you wanted to squash down. Can you squash these? Then, I am good to merge.

**Comment by hannes-ucsc** [comment]
Squashed and rebased, but had conflicts. Need to wait for test results.

**Comment by fnothaft** [comment]
@hannes-ucsc PING as integration tests failed after the rebase.

**Comment by cket** [comment]
Jenkins, test this please

**Comment by cket** [comment]
I've seen the CGCloud tests fail sporadically on pip installation before, I'll open a ticket to retry on that error

**Comment by hannes-ucsc** [comment]
I think it's more likely a concurrency issue with multiple pip instances stepping onto each other's cache toes.

**Comment by fnothaft** [comment]
This LGTM again, but do you want to squash this down further? At the least, my original intention was that all of my commits should squash down into a single commit.

**Comment by hannes-ucsc** [comment]
Ran out of buckets.

Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

**Pull request merged**